File: windows/Sources/ClearPowerCore/Ema.cs
// Exponential moving average with a wall-clock time constant.
// Port of daemon/clearpowerd/smoothing.py.
using System;

namespace ClearPower.Core
{
    public sealed class Ema
    {
        public double Tau { get; }
        public double? V { get; private set; }
        private double _t;

        public Ema(double tauS)
        {
            Tau = Math.Max(tauS, 0.0);
        }

        public double Update(double x, double t)
        {
            if (V == null || Tau == 0)
            {
                V = x;
            }
            else
            {
                var a = 1.0 - Math.Exp(-Math.Max(t - _t, 0.0) / Tau);
                V += a * (x - V.Value);
            }
            _t = t;
            return V.Value;
        }

        public void Reset() => V = null;
    }
}

File: macos/Sources/ClearPowerCore/Ema.swift
// Exponential moving average with a wall-clock time constant.
// Port of daemon/clearpowerd/smoothing.py.
import Foundation

public struct Ema {
    public let tau: Double
    public private(set) var value: Double?
    private var t: Double?

    public init(tau: Double) {
        self.tau = max(tau, 0)
    }

    @discardableResult
    public mutating func update(_ x: Double, at t: Double) -> Double {
        if let v = value, tau > 0, let t0 = self.t {
            let a = 1.0 - exp(-max(t - t0, 0) / tau)
            value = v + a * (x - v)
        } else {
            value = x
        }
        self.t = t
        return value!
    }

    public mutating func reset() {
        value = nil
    }
}
